Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,266 @@
+* TailCallHelperStack - a per thread stack of helper entries that is used to determine whether a tail call is chained or not. Its entries are allocated as local variables in CallTarget thunks. Each entry contains:
+  * Stack pointer captured right before a call to a tail call target
+  * ChainCall flag indicating whether the CallTarget thunk should return after the call to the tail call target or whether it should execute its epilog and jump to TailCallHelper instead. The latter is used by the TailCallHelper to remove the stack frame of the CallTarget before making a tail call from a tail called function. 
+  * Pointer to the next entry on the stack.
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
+* The caller executes its epilog, restoring stack pointer and callee saved registers to their values when the caller was entered.
+* The caller jumps to the TailCallHelper. This function performs the following operations:
+  * Get the topmost TailCallHelperStack entry for the current thread.
+  * Check if the previous stack frame is a CallTarget thunk frame by comparing the stack pointer value stored in the TailCallHelperStack entry to the current CFA (call frame address). If it matches, it means that the previous stack frame belongs to a CallTarget thunk and so the tail call caller was also tail called.
+  * If the previous frame was a CallTarget thunk, its stack frame needs to be removed to ensure that the stack will not grow when tail calls are chained. Set ChainCall flag in the TailCallHelperStack entry and return. That returns control to the CallTarget thunk. It checks the ChainCall flag and since it set, it executes its epilog and jumps to the TailCallHelper again.
+  * If the previous frame was not a CallTarget thunk, get the address of the CallTarget thunk of the tailcall target from the arguments buffer and jump to it.
+* The CallTarget thunk function then does the following operation:
+  * Create local instance of TailCallHelperStack entry and store the current stack pointer value in it.
+  * Push the entry to the TailCallHelperStack of the current thread.
+  * Get the arguments buffer from the thread local storage, extract the arguments, release the buffer and call the target function. The frame of the CallTarget thunk ensures that the arguments of the target are GC protected until the target function returns or tail calls to another function.
+  * Pop the TailCallHelperStack entry from the TailCallHelperStack of the current thread.
+  * Check the ChainCall flag in the TailCallHelperStack entry. If it is set, run epilog and jump to the TailCallHelper.
+  * If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function.
```

Comment:
VM indicates to JIT whether a call site requires an extra genericCtx param to be passed and also provides the context handle that is to be passed as generic context.

For the example `A() -> B()`

Say A() is getting JIT compiled and JIT is importing call to B().  JIT asks VM about `CORINFO_CALL_INFO` using `getCallInfo()` JIT-EE interface call.  CallInfo struct has `CORINFO_SIG_INFO` that provides details on signature of the call.

If  `sig->callConv` has CORINFO_CALLCONV_PARAMTYPE flag set, then such a call requires to be passed a genericCtx.  
Further callInfo struct two fields: contextHandle and whether exactContextNeedsRuntimeLookup. contextHandle encodes both the handle and a flag (CORINFO_CONTEXTFLAGS_METHOD or CORINFO_CONTEXTFLAGS_CLASS) to indicate whether the context handle is a method or class handle.  JIT derives genericCtx from contextHandle by clearing the flags.

We cannot use the same JIT-EE interface call - getCallInfo() -  for determining genericCtx that is to be passed to CallTarget IL thunk, because that interface call requires CORINFO_RESOLVED_TOKEN corresponding to the call.  While jit compiling A(), JIT has CORINFO_RESOLVED_TOKEN for call to B() and not to CallTarget IL thunk.